In [2]:
import pandas as pd

In [3]:
events = pd.read_csv("events1.csv")
items = pd.read_csv("items.csv")
users = pd.read_csv("/users.csv")

In [4]:
print(events.shape)
print(items.shape)
print(users.shape)

(758884, 7)
(1381, 6)
(270154, 3)


In [5]:
events.head()

,user_id,ga_session_id,country,device,type,item_id,date
0,2133,16909,US,mobile,purchase,94,2020-11-01 00:27:14
1,2133,16909,US,mobile,purchase,425,2020-11-01 00:27:14
2,5789,16908,SE,desktop,purchase,1,2020-11-01 01:44:44
3,5789,16908,SE,desktop,purchase,62,2020-11-01 01:44:44
4,5808,4267,US,mobile,add_to_cart,842,2020-11-01 03:06:29


In [6]:
events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 758884 entries, 0 to 758883
Data columns (total 7 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   user_id        758884 non-null  int64 
 1   ga_session_id  758884 non-null  int64 
 2   country        754329 non-null  object
 3   device         758884 non-null  object
 4   type           758884 non-null  object
 5   item_id        758884 non-null  int64 
 6   date           758884 non-null  object
dtypes: int64(3), object(4)
memory usage: 40.5+ MB


In [7]:
items.head()

,id,name,brand,variant,category,price_in_usd
0,0,Google Land & Sea Cotton Cap,Google,Single Option Only,Apparel,14
1,1,Google KeepCup,Google,Single Option Only,New,28
2,2,Google Land & Sea Nalgene Water Bottle,Google,Single Option Only,Drinkware,20
3,3,Google Unisex Eco Tee Black,Google,LG,Uncategorized Items,22
4,4,Google Chicago Campus Bottle,Google,Single Option Only,Campus Collection,11


In [8]:
items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1381 entries, 0 to 1380
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            1381 non-null   int64 
 1   name          1381 non-null   object
 2   brand         1381 non-null   object
 3   variant       973 non-null    object
 4   category      1381 non-null   object
 5   price_in_usd  1381 non-null   int64 
dtypes: int64(2), object(4)
memory usage: 64.9+ KB


In [9]:
users.head()

,id,ltv,date
0,0,0,2020-10-13 05:08:47
1,1,0,2020-11-24 14:26:54
2,2,0,2020-11-24 06:19:54
3,3,231,2020-05-02 11:09:15
4,4,102,2020-11-18 15:54:38


In [10]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 270154 entries, 0 to 270153
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   id      270154 non-null  int64 
 1   ltv     270154 non-null  int64 
 2   date    270154 non-null  object
dtypes: int64(2), object(1)
memory usage: 6.2+ MB


In [11]:
for name, df in {
    "events": events,
    "items": items,
    "users": users}.items():
    print("=" * 50)
    print(name)
    print(df.shape)
    print(df.isna().sum().sort_values(ascending=False).head(15))

events
(758884, 7)
country          4555
user_id             0
ga_session_id       0
device              0
type                0
item_id             0
date                0
dtype: int64
items
(1381, 6)
variant         408
id                0
name              0
brand             0
category          0
price_in_usd      0
dtype: int64
users
(270154, 3)
id      0
ltv     0
date    0
dtype: int64


In [12]:
print(events.columns.tolist())
print(items.columns.tolist())
print(users.columns.tolist())

['user_id', 'ga_session_id', 'country', 'device', 'type', 'item_id', 'date']
['id', 'name', 'brand', 'variant', 'category', 'price_in_usd']
['id', 'ltv', 'date']


In [61]:
events["date"] = pd.to_datetime(events["date"])

print(events["date"].min())
print(events["date"].max())

2020-11-01 00:27:14
2021-01-31 23:59:08


In [64]:
events["user_id"].nunique()

14701

In [63]:
events["ga_session_id"].nunique()

18034

In [18]:
(events["user_id"] + events["ga_session_id"]).nunique()

15434

In [13]:
events["type"].value_counts()

type
add_to_cart       667282
begin_checkout     76047
purchase           15555
Name: count, dtype: int64

In [108]:
print('event conversion: add to cart -> checkout:', '76047 / 667282 =', f"{76047 / 667282:.1%}")

event conversion: add to cart -> checkout: 76047 / 667282 = 11.4%


In [109]:
print('event conversion: checkout -> purchase:', '15555 / 76047 =', f"{15555 / 76047:.1%}")

event conversion: checkout -> purchase: 15555 / 76047 = 20.5%


In [110]:
print('event conversion: add to cart -> purchase:', '15555 / 667282 =', f"{15555 / 667282:.1%}")

event conversion: add to cart -> purchase: 15555 / 667282 = 2.3%


In [107]:
funnel_users = (
    fact.groupby("type")
    ["user_id"]
    .nunique()
    .reset_index(name="users")
)

funnel_users

,type,users
0,add_to_cart,12545
1,begin_checkout,6404
2,purchase,4066


In [111]:
print('user conversion: add to cart -> checkout:', '6404 / 12545 =', f"{6404 / 12545:.1%}")

user conversion: add to cart -> checkout: 6404 / 12545 = 51.0%


In [114]:
print('user conversion: checkout -> purchase:', '4066 / 6404 =', f"{4066 / 6404:.1%}")

user conversion: checkout -> purchase: 4066 / 6404 = 63.5%


In [115]:
print('user conversion: add to cart -> purchase:', '4066 / 12545 =', f"{4066 / 12545:.1%}")

user conversion: add to cart -> purchase: 4066 / 12545 = 32.4%


The largest drop-off occurs between Add to Cart and Checkout stages. 
Only 51% of users proceed to checkout, while 63.5% of checkout users complete a purchase. This suggests that optimization efforts should focus on cart experience, checkout entry points and purchase intent activation

<h3>marts for dash</h3>

In [37]:
fact = (events.merge(
        items,
        left_on="item_id",
        right_on="id",
        how="left"))

In [50]:
print(fact.shape)
print(fact.info())
fact.head()

(758884, 13)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 758884 entries, 0 to 758883
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   user_id        758884 non-null  int64 
 1   ga_session_id  758884 non-null  int64 
 2   country        754329 non-null  object
 3   device         758884 non-null  object
 4   type           758884 non-null  object
 5   item_id        758884 non-null  int64 
 6   date           758884 non-null  object
 7   id             758884 non-null  int64 
 8   name           758884 non-null  object
 9   brand          758884 non-null  object
 10  variant        122624 non-null  object
 11  category       758884 non-null  object
 12  price_in_usd   758884 non-null  int64 
dtypes: int64(5), object(8)
memory usage: 75.3+ MB
None


,user_id,ga_session_id,country,device,type,item_id,date,id,name,brand,variant,category,price_in_usd
0,2133,16909,US,mobile,purchase,94,2020-11-01 00:27:14,94,Google Large Tote White,Google,Single Option Only,Bags,10
1,2133,16909,US,mobile,purchase,425,2020-11-01 00:27:14,425,Google Heather Green Speckled Tee,Google,XL,Apparel,21
2,5789,16908,SE,desktop,purchase,1,2020-11-01 01:44:44,1,Google KeepCup,Google,Single Option Only,New,28
3,5789,16908,SE,desktop,purchase,62,2020-11-01 01:44:44,62,Google Mini Kick Ball,Google,Single Option Only,Fun,2
4,5808,4267,US,mobile,add_to_cart,842,2020-11-01 03:06:29,842,Google Infant Hero Onesie Grey,Google,NaN,Apparel,25


In [47]:
fact["name"].isna().sum()

0

In [79]:
fact["date"] = pd.to_datetime(fact["date"])

In [80]:
fact["date"] = pd.to_datetime(fact["date"])

In [118]:
fact["device"].value_counts()

device
desktop    441521
mobile     300628
tablet      16735
Name: count, dtype: int64

<h3>metrics</h3>

In [81]:
purchases = fact[fact["type"] == "purchase"]

In [82]:
purchases["price_in_usd"].sum()

307114

In [83]:
len(purchases)

15555

In [116]:
round(purchases["price_in_usd"].mean(), 2)

19.74

In [95]:
fact["event_date"] = fact["date"].dt.date

In [96]:
dau = (
    fact.groupby("event_date")
    ["user_id"]
    .nunique()
    .reset_index(name="dau")
)

In [97]:
dau.head()

,event_date,dau
0,2020-11-01,14
1,2020-11-02,39
2,2020-11-03,46
3,2020-11-04,30
4,2020-11-05,26


In [101]:
revenue_daily = (
    purchases
    .groupby("event_date")
    ["price_in_usd"]
    .sum()
    .reset_index(name="revenue")
)

In [102]:
revenue_daily.head()

,event_date,revenue
0,2020-11-01,648
1,2020-11-02,3553
2,2020-11-03,2874
3,2020-11-04,2010
4,2020-11-05,1433


In [89]:
event_summary = (
    fact.groupby("type")
    .agg(
        events=("user_id","count"),
        users=("user_id","nunique")
    )
    .reset_index()
)

event_summary

,type,events,users
0,add_to_cart,667282,12545
1,begin_checkout,76047,6404
2,purchase,15555,4066


In [117]:
category_revenue = (
    purchases
    .groupby("category")
    ["price_in_usd"]
    .sum()
    .sort_values(ascending=False)
    .reset_index()
)

category_revenue.head(15)

,category,price_in_usd
0,Apparel,167316
1,Bags,19707
2,New,19179
3,Campus Collection,16260
4,Shop by Brand,15473
5,Accessories,14637
6,Drinkware,11008
7,Uncategorized Items,10967
8,Lifestyle,10366
9,Clearance,8949


<h3>finals</h3>

In [119]:
fact = fact.merge(
    users,
    left_on="user_id",
    right_on="id",
    how="left",
    suffixes=("", "_user")
)

In [120]:
fact.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 758884 entries, 0 to 758883
Data columns (total 17 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   user_id        758884 non-null  int64         
 1   ga_session_id  758884 non-null  int64         
 2   country        754329 non-null  object        
 3   device         758884 non-null  object        
 4   type           758884 non-null  object        
 5   item_id        758884 non-null  int64         
 6   date           758884 non-null  datetime64[ns]
 7   id             758884 non-null  int64         
 8   name           758884 non-null  object        
 9   brand          758884 non-null  object        
 10  variant        122624 non-null  object        
 11  category       758884 non-null  object        
 12  price_in_usd   758884 non-null  int64         
 13  event_date     758884 non-null  object        
 14  id_user        758884 non-null  int64         
 15  

In [130]:
fact["event_date"] = pd.to_datetime(fact["event_date"])

In [131]:
fact.to_csv(
    "fact_product_analytics.csv",
    index=False
)

In [132]:
fact.to_parquet(
    "fact_product_analytics.parquet",
    index=False
)